scipy/stats/morestats.py

```diff
@@ -1340,6 +1340,10 @@ def yeojohnson(x, lmbda=None):
     if x.size == 0:
         return x
 
+    if not np.issubdtype(x.dtype, np.floating):
```

Comment:
There are methods used in scipy where gradients are calculated from complex inputs. If you'd like to instead write a unit test that confirms those values that would be helpful. I don't think it is necessary to restrict to only float types here.